In [2]:
import sys
import io

# sys.stdout = io.TextIOWrapper(sys.stdout.buffer, encoding='utf-8')

import sentence_transformers
import faiss
import PyPDF2

print("Môi trường RAG đã sẵn sàng hoạt động tuyệt vời!")
print(f"Phiên bản Sentence-Transformers: {sentence_transformers.__version__}")
from sentence_transformers import SentenceTransformer
import numpy as np
import torch 
import torch.nn.functional as F


d:\anaconda\envs\rag_env\lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


Môi trường RAG đã sẵn sàng hoạt động tuyệt vời!
Phiên bản Sentence-Transformers: 6.1.0


In [5]:
def cosine_similarity(v1, v2):
    dot = np.dot(v1, v2)
    norm = np.linalg.norm(v1)* np.linalg.norm(v2)
    return dot / norm if norm != 0 else 0.0


if __name__ == "__main__":
    # v1 = np.array([1, 2, 3])
    # v2 = np.array([4, 5, 6])
    # similarity = cosine_similarity(v1, v2)
    # print(f"Cosine similarity between {v1} and {v2} is: {similarity}")
    model  = SentenceTransformer("all-MiniLM-L6-v2")
    sentences = [' the cat is running', "the weather is bad", "we can run on the field",
                "do you love me forever", "i really love you my son", "this book is so boring", 
                " Watch out !", "I am Steve", "The dogs are hungry", "don't go to swim today !"]
    embeddings = model.encode(sentences)
    print(embeddings.shape)
    embed_tensor = torch.tensor(embeddings)
    normalize_v = F.normalize(embed_tensor, dim = 1)
    cosine_sim_matrix  = torch.mm(normalize_v, normalize_v.t())
    print(f"Do tuong dong cua 10 cau sau khi embedding sentences : {cosine_sim_matrix}")
    

Loading weights: 100%|██████████| 103/103 [00:00<00:00, 2360.37it/s]


(10, 384)
Do tuong dong cua 10 cau sau khi embedding sentences : tensor([[ 1.0000, -0.0719,  0.1820,  0.0759,  0.0561,  0.0476, -0.0329, -0.0204,
          0.1878,  0.0116],
        [-0.0719,  1.0000,  0.1226,  0.0173,  0.0482,  0.1343,  0.0992,  0.0925,
         -0.0424,  0.3055],
        [ 0.1820,  0.1226,  1.0000,  0.1851,  0.1339,  0.0618,  0.0400,  0.1210,
         -0.0485,  0.0963],
        [ 0.0759,  0.0173,  0.1851,  1.0000,  0.4710,  0.0585,  0.1253,  0.1921,
          0.0128,  0.1364],
        [ 0.0561,  0.0482,  0.1339,  0.4710,  1.0000,  0.0357,  0.1316,  0.1928,
          0.0224,  0.1229],
        [ 0.0476,  0.1343,  0.0618,  0.0585,  0.0357,  1.0000,  0.1384,  0.0681,
          0.0598,  0.1396],
        [-0.0329,  0.0992,  0.0400,  0.1253,  0.1316,  0.1384,  1.0000,  0.1453,
          0.0074,  0.3782],
        [-0.0204,  0.0925,  0.1210,  0.1921,  0.1928,  0.0681,  0.1453,  1.0000,
         -0.0844,  0.1544],
        [ 0.1878, -0.0424, -0.0485,  0.0128,  0.0224,  0.0598, 

In [6]:
print(cosine_sim_matrix.shape)

torch.Size([10, 10])


In [2]:
import os
folders = [
    "data/raw_docs", 
    "data/vector_store", 
    "src",
    "scripts",
    "notebook",

]
for folder in folders: 
    os.makedirs(folder, exist_ok = True)
    if folder == "src": 
        init_file = os.path.join(folder, "__init__.py")
        if not os.path.exists(init_file):
            open(init_file, "a").close

print(f"Khoi tao thanh cong tree Folders cho Project Chatbox RAG ")



Khoi tao thanh cong tree Folders cho Project Chatbox RAG 


In [4]:
pip install datasets

   ---------------------------------------- 0.0/2.2 MB ? eta -:--:--
   --------- ------------------------------ 0.5/2.2 MB 5.6 MB/s eta 0:00:01
   --------------------------------- ------ 1.8/2.2 MB 9.1 MB/s eta 0:00:01
   --------------------------------- ------ 1.8/2.2 MB 9.1 MB/s eta 0:00:01
   --------------------------------- ------ 1.8/2.2 MB 9.1 MB/s eta 0:00:01
   ---------------------------------------- 2.2/2.2 MB 2.1 MB/s  0:00:01

   ---------------------------------------- 0/5 [Mako]
   ---------------------------------------- 0/5 [Mako]
   ---------------------------------------- 0/5 [Mako]
   ---------------------------------------- 0/5 [Mako]
   -------- ------------------------------- 1/5 [greenlet]
   -------- ------------------------------- 1/5 [greenlet]
   -------- ------------------------------- 1/5 [greenlet]
   -------- ------------------------------- 1/5 [greenlet]
   -------- ------------------------------- 1/5 [greenlet]
   -------- --------------------------

# Problem : XÂY DỰNG HỆ THỐNG CHATBOX KẾT HỢP RAG ĐƯA RA CÂU TRẢ LỜI ĐÚNG/ SAI & GIẢI THÍCH TỪ CÁC READING PASSAGES.

## Objective: 
+ Có thêm kiến thức LLMs, RAG, systems design ML/ DL. 
+ Giúp cho LLMs model cải thiện/ hạn chế tình trạng mơ hồ.
+ Cải thiện độ chính xác của câu trả lời, đưa ra giải thích hợp lí.

**Hướng tiếp cận :**
* Giai đoạn 1 (Khuyên dùng): Đóng băng (Freeze) LLM hoàn toàn. Dùng kỹ thuật In-context Learning và Prompt Engineering (hướng dẫn LLM đóng vai trợ lý sửa bài đọc hiểu dựa trên context được retrieve).   

* Giai đoạn 2 (Mở rộng nếu muốn học sâu về DL): Thay vì fine-tune LLM sinh văn bản, bạn có thể thử nghiệm fine-tune mô hình Embedding/Retriever (như Sentence Transformers) hoặc thử nghiệm fine-tune nhẹ Generator bằng LoRA/QLoRA để chuẩn hóa định dạng câu trả lời sư phạm

## Dataset :

RACE (https://www.cs.cmu.edu/~glai1/data/race/):
+ Mô tả: 

    Quy mô & Thách thức: Gồm 27,933 bài đọc và 97,687 câu hỏi. Khoảng cách lớn giữa mô hình SOTA lúc bấy giờ (43%) và ngưỡng con người (95%), đòi hỏi khả năng suy luận (reasoning) phức tạp.

### Data Analysis 

**Thống kê dataset:** Chia thành RACE-M (THCS: 25,421 câu hỏi) và RACE-H (THPT: 62,445 câu hỏi).Bài đọc của RACE-H dài hơn (trung bình 353.1 từ) và từ vựng phong phú hơn (125,120 từ) so với RACE-M
Phân loại mức độ suy luận (Reasoning Types):
    Word matching (Khớp từ khóa): 15.8%
    Paraphrasing (Diễn giải lại): 19.2%
    Single-sentence reasoning (Suy luận đơn câu): 33.4%
    Multi-sentence reasoning (Suy luận đa câu): 25.8%
    Insufficient/Ambiguous (Mơ hồ): 5.8%

Tổng tỷ lệ câu hỏi cần suy luận đạt 59.2%, cao hơn vượt trội so với CNN (21%) hay SQuAD (20.5%).

Phân loại chi tiết dạng câu hỏi: Bao gồm suy luận chi tiết (detail reasoning), hiểu toàn cảnh (whole-picture reasoning), tóm tắt bài đọc (passage summarization), phân tích thái độ tác giả (attitude analysis) và tri thức thế giới (world knowledge)

**Collection Methodology**
Thu thập từ 3 website đề thi công cộng lớn
Quy trình lọc 3 bước: Chỉ giữ câu hỏi trắc nghiệm 4 lựa chọn; loại bỏ các câu hỏi chứa hình ảnh/bảng biểu hoặc tham chiếu gạch chân; loại bỏ trùng lặp và dùng phần mềm OCR (Tesseract / ABBYY FineReader) để chuyển đổi đáp án dạng ảnh.

## PIPELINE FOR RAG
<img src="RAG_pipeline_img.webp" width="450"/>

### INGESTION PIPELINE

RACE JSON

   ↓ 1. LOAD & PARSE     → extract article, question, options, answer

   ↓ 2. CHUNK            → chia article thành đoạn nhỏ có nghĩa
   
   ↓ 3. EMBED            → chuyển text thành vector số
   
   ↓ 4. INDEX            → lưu vector vào FAISS, gắn metadata
   
        Vector DB sẵn sàng để retrieve


**References:**
- Retrieval-Augmented Generation for Knowledge-Intensive NLP Tasks (Patrick Lewis†‡, Ethan Perez ,Aleksandra Piktus†, Fabio Petroni†,Vladimir Karpukhin†, Naman Goyal†, Heinrich Küttler†, Mike Lewis†, Wen-tau Yih†, Tim Rocktäschel†‡, Sebastian Riedel†‡, Douwe Kiela) [https://arxiv.org/pdf/2005.11401]

- RACE: Large-scale ReAding Comprehension Dataset From Examinations (Guokun Lai∗and Qizhe Xie∗and Hanxiao Liu and Yiming Yang and Eduard Hovy) [https://arxiv.org/pdf/1704.04683]

- Retrieval-Augmented Generation (RAG) [https://www.pinecone.io/learn/retrieval-augmented-generation/]


## INGESTION DOCUMENTS

### 1. LOAD & PARSE DATA
-hugging face [https://huggingface.co/docs/datasets/load_hub]
### 2. CHUNKING
- **StackOverflow Blog: Breaking up is hard to do** [https://stackoverflow.blog/2024/12/27/breaking-up-is-hard-to-do-chunking-in-rag-applications/]

    **2.1. Why is chunking important?**

        + Chunking quá lớn -> mất khả năng cụ thể 

        + chunking quá ít -> mất bối cảnh.

    **=> Kích thước phải phù hợp**

        + kích thước của chunk quan trọng, ngoài ra kích thước của query vector cần phải matching với chunking vector. 

        + metadata là 1 cách để liên kết tới phần gốc của văn bản.

    **2.2 Chunking strategies**
        + Cách cơ bản : chia theo fixed-size. (phương pháp rẻ nhất về lượng tính toán ta cần, nhưng nó không tính đến ngữ cảnh của nội dung đang chia nhỏ. )

        + random chunk sizes : gồm nhiều loại tài liệu, do nhiều bối cảnh, chủ đề (không nên làm do tính ngẫu nhiên )
    
        + sliding windows methods : đối với document thứ n+1 sẽ overlap lại document thứ n và obtain thêm document của chính n+1.

        + Context-aware chunking: chia dựa trên dấu câu (yêu cầu tính toán, **document thay đổi theo thời gian gặp struggle**)  

        + adaptive chunking: Dùng ML để chia thành các unit nhỏ hơn có context tốt, highly-tailored(computation cao)


    **Kết luận** : Khi bạn kiểm tra các phương pháp chia đôi, hãy kiểm tra kết quả của hệ thống RAG với các truy vấn mẫu. Đánh giá họ bằng đánh giá của con người và với các chuyên gia đánh giá LLM. Khi bạn đã xác định được phương pháp nào luôn hiệu quả hơn, bạn có thể nâng cao kết quả bằng cách lọc kết quả dựa trên điểm tương đồng cosine.







_____________________________________________________________________________________________

- **5 Levels of Text Splitting (Notebook)** [https://github.com/FullStackRetrieval-com/RetrievalTutorials/blob/main/tutorials/LevelsOfTextSplitting/5_Levels_Of_Text_Splitting.ipynb]


    | Cấp độ  | Ý tưởng chính                                   |
    | ------- | ----------------------------------------------- |
    | Level 1 | Chia theo số ký tự cố định                      |
    | Level 2 | Chia theo các dấu phân tách                     |
    | Level 3 | Chia theo cấu trúc của từng loại tài liệu       |
    | Level 4 | Chia theo ngữ nghĩa bằng Embedding              |
    | Level 5 | Để AI tự quyết định cách chia                   |
    | Bonus   | Tạo thêm biểu diễn dữ liệu để retrieval tốt hơn |


    level 1: 
        + có thể dùng **from langchain.text_splitter import CharacterTextSplitter** để split + overlap theo số lượng từ, sau đó bộc trong documents.
        + llama index: 

    level 2:

        + có thể dùng **from langchain.text_splitter import RecursiveCharacterTextSplitter**: Chia theo đoạn văn bản nếu có xuống dòng, dấu câu chẳng hạn. Qua ảnh trực quan hóa nó khả tốt với tài liệu văn bản ta hướng tới.
 
    level 3: Document Specific Splitting - không cần lắm do tài liệu đẫ đc trích xuất oke oke 

    level 4;
    level 5; 

__________________________________________________________________________________________________________________________________________

- **Unstructured.io: Chunking for RAG best practices** [unstructured.io/blog/chunking-for-rag-best-practices]



### 3. EMBEDDINGS
- **Sentence-BERT paper (Reimers & Gurevych 2019)** : [arxiv.org/abs/1908.10084]
    
    Sentence BERT được build để tăng tốc độ tính toán khi embedding cả câu, thay vì từng character như BERT based.
     
- **SBERT docs — Pretrained Models** : [sbert.net/docs/sentence_transformer/pretrained_models.html]

- **Blog: The Illustrated Word2Vec (Jay Alammar)** : [jalammar.github.io/illustrated-word2vec]


### 4. FAISS& VECTOR DB 

